src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,246 @@
+    return false;
+  }
+  bool minimum_time_reached(const ceph::timespan& curr_timestamp, const hash_map::iterator& it)
+  {
+    using namespace std::chrono;
+    constexpr auto min_duration = duration_cast<ceph::timespan>(seconds(60)) / fixed_float_rgw_ratelimit;
+    const auto delta = curr_timestamp - it->second.ts;
+    if (delta < min_duration)
+      return false;
+    return true;
+  }
+  bool increase_tokens(hash_map::iterator& it, const ceph::timespan& curr_timestamp,
+                       const RGWQoSInfo& tenant)
+  {
+    constexpr int fixed_float = fixed_float_rgw_ratelimit;
+    if(curr_timestamp > it->second.ts && !it->second.first_run && minimum_time_reached(curr_timestamp, it))
+    {
+      int64_t time_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(curr_timestamp - it->second.ts).count() / 60.0 / std::milli::den * fixed_float; // / 60 to make it work with 1 min token bucket
+      it->second.ts = curr_timestamp;
+      int64_t write_ops = tenant.max_write_ops * time_in_ms;
+      int64_t write_bw = tenant.max_write_bytes * time_in_ms;
+      int64_t read_ops = tenant.max_read_ops * time_in_ms;
+      int64_t read_bw = tenant.max_read_bytes * time_in_ms;
+      it->second.read[RGW_QOS_OPS] = std::min(tenant.max_read_ops * fixed_float, read_ops + it->second.read[RGW_QOS_OPS]);
+      it->second.read[RGW_QOS_BW] = std::min(tenant.max_read_bytes * fixed_float, read_bw + it->second.read[RGW_QOS_BW]);
+
+      it->second.write[RGW_QOS_OPS] = std::min(tenant.max_write_ops * fixed_float, write_ops + it->second.write[RGW_QOS_OPS]);
+      it->second.write[RGW_QOS_BW] = std::min(tenant.max_write_bytes * fixed_float, write_bw + it->second.write[RGW_QOS_BW]);
+
+    } else if(it->second.first_run)
+      {
+        it->second.write[RGW_QOS_OPS] = tenant.max_write_ops * fixed_float;
+        it->second.write[RGW_QOS_BW] = tenant.max_write_bytes * fixed_float;
+        it->second.read[RGW_QOS_OPS] = tenant.max_read_ops * fixed_float;
+        it->second.read[RGW_QOS_BW] = tenant.max_read_bytes * fixed_float;
+        it->second.ts = curr_timestamp;
+        it->second.first_run = false;
+      }
+      return true;
+    }
+
+  bool increase_entry_write(hash_map::iterator& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((it->second.write_ops() - 1 < 0) && (ops_limit > 0)) ||
+      (it->second.write_bytes() < 0 && bw_limit > 0))
+    {
+      return true;
+    }
+
+    // we don't want to reduce ops' tokens if we've rejected it.
+    it->second.write[RGW_QOS_OPS] -= fixed_float_rgw_ratelimit;
+    return false;
+  }
+
+  public:
+    QosDatastruct(const QosDatastruct&) = delete;
+    QosDatastruct& operator =(const QosDatastruct&) = delete;
+    QosDatastruct(QosDatastruct&&) = delete;
+    QosDatastruct& operator =(QosDatastruct&&) = delete;
+    QosDatastruct() {};
```

Comment:
should not have a default constructor. this would leave `replacing` and `cv` uninitialized